src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiTask.cs

```diff
@@ -0,0 +1,392 @@
+        };
+
+        double max = Database.Configuration.Ai.EmbeddingsGenerationMaxFallbackTime.AsTimeSpan.TotalSeconds;
+        FallbackTime = TimeSpan.FromSeconds(Math.Min(secondsToWait, max));
+    }
+
+    protected override int LoadInternal(IEnumerable<GenAiScriptResult> items, DocumentsOperationContext context, GenAiStatsScope scope)
+    {
+        var results = PrepareItemsBeforeSendingToModel(items);
+        if (results.Count is 0)
+            return 0;
+
+        var exceptions = SendToModel(results, context);
+
+        ApplyUpdateScript(context, results);
+
+        if (exceptions is not null)
+            throw new AggregateException(exceptions);
+
+        return results.Count;
+    }
+
+    private void ApplyUpdateScript(DocumentsOperationContext context, List<GenAiResultItem> results)
+    {
+        PatchRequest req = new(Configuration.Update, PatchRequestType.AiGen);
+        var cmd = new GenAiBatchPatchCommand(context, results, req, Configuration.Name);
+
+        Database.TxMerger.Enqueue(cmd).GetAwaiter().GetResult();
+    }
+
+    protected override GenAiStatsScope CreateScope(EtlRunStats stats)
+    {
+        return new GenAiStatsScope(stats);
+    }
+
+    protected override string StatsAggregatorTag => "Generative AI";
+
+    protected override bool ShouldFilterOutHiLoDocument()
+    {
+        return true;
+    }
+
+    private List<Exception> SendToModel(List<GenAiResultItem> items, DocumentsOperationContext context)
+    {
+        context.CloseTransaction();
+
+        List<Exception> exceptions = null;
+        List<Task<(string Result, string Usage)>> tasks = [];
+
+        foreach (var item in items)
+        {
+            if (item.ContextOutput.IsCached)
+                continue; // no change, can skip
+            
+            string json = item.ContextOutput.Context.ToString();
+            tasks.Add(_chatCompletionClient.CompleteAsync(Configuration.Prompt, json));
+        }
+
+        try
+        {
+            // TODO: Yuck
+            Task.WaitAll(tasks.OfType<Task>().ToArray());
+        }
+        catch
+        {
+            // we'll handle that later
+        }
+
+        for (int index = 0; index < tasks.Count; index++)
+        {
+            var task = tasks[index];
+            var item = items[index];
+            if (task.IsCompletedSuccessfully is false)
+            {
+                exceptions ??= [];
+                exceptions.Add(task.Exception);
+
+                continue; // so we won't try to save it 
+            }
+
+            (string result, string usage) = task.Result;
+            // TODO: report usage
+
+            //TODO: REALLY YUCKY!
+            var stream = new ReadOnlyMemoryStream<byte>(Encoding.UTF8.GetBytes(result));
+            item.ModelOutput = new ModelOutput
+            {
+                Output = context.ReadForMemoryAsync(stream, item.DocId).GetAwaiter().GetResult()
```

Comment:
Please check `ValueTask instances should not have their result directly accessed unless the instance has already completed` warning here